src/Effectful/HPQTypes.hs

```diff
@@ -71,72 +79,128 @@ runEffectDB ::
     runWithState eff =
       PQ.withConnection (PQ.unConnectionSource connectionSource) $ \conn -> do
-        let dbState0 = mkDBState conn transactionSettings
-        evalState dbState0 eff :: Eff es a
-    runWithNewConnection ::
-      Eff (State (PQ.DBState (Eff es)) : es) b ->
-      Eff (State (PQ.DBState (Eff es)) : es) b
-    runWithNewConnection action = do
-      dbState :: PQ.DBState (Eff es) <- get
-      result <- PQ.withConnection (PQ.unConnectionSource connectionSource) $ \newConn -> do
-        -- TODO: We do not pass the current connection source to the new DB
-        -- state, which differs from the original code.
-        put $ mkDBState newConn (PQ.dbTransactionSettings dbState)
-        action
-      put dbState
-      pure result
-    mkDBState :: PQ.Connection -> PQ.TransactionSettings -> PQ.DBState (Eff es)
-    mkDBState conn ts =
-      PQ.DBState
-        { PQ.dbConnection = conn
-        , PQ.dbConnectionSource = PQ.unConnectionSource connectionSource
-        , PQ.dbTransactionSettings = ts
-        , PQ.dbLastQuery = PQ.SomeSQL (mempty :: PQ.SQL)
-        , PQ.dbRecordLastQuery = True
-        , PQ.dbQueryResult = Nothing
-        }
+        let dbState0 = mkDBState (PQ.unConnectionSource connectionSource) conn transactionSettings
+        evalState dbState0 $ handleAutoTransaction transactionSettings withTransaction' eff
+    withTransaction' ::
+      PQ.TransactionSettings ->
+      Eff (DBState es : es) a ->
+      Eff (DBState es : es) a
+    withTransaction' ts eff = unDBEff . PQ.withTransaction' ts $ DBEff eff
+
+mkDBState ::
+  PQ.ConnectionSourceM m ->
+  PQ.Connection ->
+  PQ.TransactionSettings ->
+  PQ.DBState m
+mkDBState connectionSource conn ts =
+  PQ.DBState
+    { PQ.dbConnection = conn
+    , PQ.dbConnectionSource = connectionSource
+    , PQ.dbTransactionSettings = ts
+    , PQ.dbLastQuery = PQ.SomeSQL (mempty :: PQ.SQL)
+    , PQ.dbRecordLastQuery = True
+    , PQ.dbQueryResult = Nothing
+    }
+
+handleAutoTransaction ::
+  PQ.TransactionSettings ->
+  (PQ.TransactionSettings -> m a -> m a) ->
+  m a ->
+  m a
+handleAutoTransaction transactionSettings withTransaction action =
+  -- REVIEW: Why don't we have to set `tsAutoTransaction` to `False` in the
+  -- context of the `action`?
+  if PQ.tsAutoTransaction transactionSettings
+    then withTransaction (transactionSettings {PQ.tsAutoTransaction = False}) action
+    else action
+
+---------------------------------------------------
+-- Internal effect stack
+---------------------------------------------------
+
+-- | Newtype wrapper over the internal DB effect stack
+newtype DBEff es a = DBEff
+  { unDBEff :: Eff (DBState es : es) a
+  }
+  deriving newtype (Functor, Applicative, Monad, MonadThrow, MonadCatch, MonadMask)
+
+-- | Internal state effect used to reinterpret the `DB` effect
+type DBState es = State (PQ.DBState (DBEff es))
+
+-- Convenience @MonadBase IO@ instance
+instance (IOE :> es) => MonadBase IO (DBEff es) where
```

Comment:
Btw, `MonadIO` should be preferred to `MonadBase IO` since `MonadBase` turned out to be an unnecessary generalization and when you assemble multiple libraries, you'll pull in code that uses `liftIO`anyway. Also, the former is generally "better" in a sense that it's a canonical way to lift IO actions (and is in `base`).

I'm aware `hpqtypes` itself uses `MonadBase IO`, but I'll change that in the near future (I need to sort out the mess with `ConnectionSourceM` and make the `DBState` to not be parametrized by the monad), in the meantime let's start phasing out `MonadBase` already.

Reply:
Ah, I didn't know that `MonadIO` should be preferred.  I added the corresponding instance.

I looks like we still need the `MonadBase IO` instance though, because of functions like [`simpleSource`](https://hackage.haskell.org/package/hpqtypes-1.9.4.0/docs/Database-PostgreSQL-PQTypes-Internal-Connection.html#v:simpleSource) and I suppose in general because `hpqtypes` itself uses `MonadBase IO` as you said.

I added a comment that this instance is provisional and to be removed once `hpqtypes` stops using `MonadBase IO`.

Reply:
Why not removing the instance and forcing the users to use `liftIO` downstream (this way you reduce the need to change the code when hpqtypes gets away from`MonadBase IO` I think?

Reply:
My change made it possible to be removed.